Add support for per-serial feature querying

X-Ryl669 opened the pull request:
### Description

Now that we have per-serial features (like Meatpack, and I hope, binary transfers in the future), it doesn't make sense to answer Meatpack on a serial port that does not support it in M115.

This adds a `features()` method to the serial class that's computing a features' mask that's then checked when answering M115 for the serial port that's being queried upon. 
(Honestly, if you read the previous sentence without restarting from the beginning at least 2 times, you're a champion).

### Requirements

None

### Benefits

Although it's currently useless since most host don't care about the returned features, at least, Marlin should tell the truth and only claim the features it's actually supporting.

### Configurations

N/A

### Related Issues

None... Yet.

X-Ryl669 committed:
Add support for per-serial feature querying

X-Ryl669 commented:
@thisiskeithb You're fast!

thisiskeithb commented:
> You're fast!

I just happened to be at the computer & in the issue queue at the exact same time 🙂

X-Ryl669 commented:
@rhapsodyv Can you try if it works on your dual terminal bench, please ? I don't have a board with 2 (external) serials. 
It should only report MEATPACK on the serial port where it's enabled.

thinkyhead committed:
Update macros.h

thinkyhead committed:
Update serial_base.h

thinkyhead commented:
As part of these additions, feel free to extend the "`Dnnn` Debug G-codes" or to add some "dev mode only" options to existing G-codes so that we can ask Marlin things like "what is the current verbose description of serial port 0?" Those kinds of extras can be useful in a pinch. The leveling debugging is probably an example of overkill, but it has helped narrow down issues more than a few times.

AFAIK no other firmwares currently have standard codes for inquiring into device states (though we can retrieve or change the current baud rate) and we probably don't need to formalize them (until the HAL becomes more like MF DOS) so it's fine to be informal about these and focus on what helps most with our ongoing debugging needs.

X-Ryl669 commented:
MF DOS: Marlin's Fantastic Debugging Operating System ?

X-Ryl669 committed:
Add support for dumping the current serial configuration on the serial port

X-Ryl669 commented:
Added D7 to dump the current serial configuration. It's dumping the type of the serial implementation as text (so it's the true object used, not something deduced from a chain of supposition and/or defines).

I've updated the BUG_ON macro because it was failing on AVR (due to SERIAL_ECHOPAIR requiring the first argument to be a PSTR/pgm string, which is not the case of a `const char*`). It's a little bit more verbose but it build.

I've added a template `gtn` function that's returning the type of the parameter used as string **without** relying on RTTI, so it's useful for debugging (like for the gcode above). It's inflating the binary ~27 bytes per call, so use sparsely.

X-Ryl669 committed:
Add buffer size to the dump

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/21318

thinkyhead committed:
post-merge tweaks

lilinormalgirl commented:
Wie funktioniert musk.el1_2 ohne Versuch um Date runter zu downloaden

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/21318

thinkyhead committed:
space

thisiskeithb commented:
I'm now getting a huge mess of warnings after merging this PR. Link to one of my forks with this issue: [boop](https://github.com/thisiskeithb/Marlin/tree/myconfig/MK3S-BTT002-TMC2209).

<details><summary>Error details:</summary>
<p>

```prolog
                 from Marlin/src/gcode/probe/G30.cpp:23:
Marlin/src/gcode/probe/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/probe/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
In file included from Marlin/src/gcode/probe/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/probe/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/probe/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/probe/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/probe/../../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/probe/M401_M402.cpp:23:
Marlin/src/gcode/probe/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/probe/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
In file included from Marlin/src/gcode/probe/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/probe/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/probe/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/probe/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/probe/../../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/probe/M851.cpp:23:
Marlin/src/gcode/probe/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/probe/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
Compiling .pio/build/BIGTREE_BTT002/src/src/gcode/stats/M31.cpp.o
Compiling .pio/build/BIGTREE_BTT002/src/src/gcode/stats/M75-M78.cpp.o
In file included from Marlin/src/gcode/../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/queue.h:29,
                 from Marlin/src/gcode/queue.cpp:27:
Marlin/src/gcode/../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
In file included from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/sd/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/sd/../../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/sd/M1001.cpp:23:
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
In file included from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/sd/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/sd/../../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/sd/M20.cpp:23:
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
Compiling .pio/build/BIGTREE_BTT002/src/src/gcode/temp/M104_M109.cpp.o
In file included from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/sd/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/sd/../../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/sd/M21_M22.cpp:23:
Compiling .pio/build/BIGTREE_BTT002/src/src/gcode/temp/M105.cpp.o
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
In file included from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/sd/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/sd/../../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/sd/M23.cpp:23:
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
Compiling .pio/build/BIGTREE_BTT002/src/src/gcode/temp/M106_M107.cpp.o
In file included from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/sd/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/sd/../../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/sd/M24_M25.cpp:23:
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
Compiling .pio/build/BIGTREE_BTT002/src/src/gcode/temp/M140_M190.cpp.o
In file included from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/MarlinSerial.h:27,
Compiling .pio/build/BIGTREE_BTT002/src/src/gcode/temp/M141_M191.cpp.o
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/sd/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/sd/../../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/sd/M26.cpp:23:
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
Compiling .pio/build/BIGTREE_BTT002/src/src/gcode/temp/M155.cpp.o
In file included from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/sd/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/sd/../../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/sd/M27.cpp:23:
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
In file included from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/sd/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/sd/../../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/sd/M28_M29.cpp:23:
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
Compiling .pio/build/BIGTREE_BTT002/src/src/gcode/temp/M303.cpp.o
In file included from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/sd/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/sd/../../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/sd/M30.cpp:23:
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
In file included from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/sd/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/sd/../../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/sd/M32.cpp:23:
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
In file included from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/sd/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/sd/../../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/sd/M34.cpp:23:
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
In file included from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/sd/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/sd/../../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/sd/M33.cpp:23:
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
Compiling .pio/build/BIGTREE_BTT002/src/src/gcode/units/M82_M83.cpp.o
In file included from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/sd/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/sd/../../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/sd/M524.cpp:23:
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
Compiling .pio/build/BIGTREE_BTT002/src/src/lcd/HD44780/lcdprint_hd44780.cpp.o
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
In file included from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/sd/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/sd/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/sd/../../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/sd/M928.cpp:23:
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/sd/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
Compiling .pio/build/BIGTREE_BTT002/src/src/lcd/HD44780/marlinui_HD44780.cpp.o
Compiling .pio/build/BIGTREE_BTT002/src/src/lcd/extui/ui_api.cpp.o
Compiling .pio/build/BIGTREE_BTT002/src/src/lcd/fontutils.cpp.o
In file included from Marlin/src/gcode/stats/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/stats/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/stats/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/stats/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/stats/../../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/stats/../gcode.h:303,
                 from Marlin/src/gcode/stats/M31.cpp:23:
Marlin/src/gcode/stats/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/stats/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
Compiling .pio/build/BIGTREE_BTT002/src/src/lcd/lcdprint.cpp.o
In file included from Marlin/src/gcode/stats/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/stats/../../inc/../HAL/./STM32/MarlinSerial.h:27,
Compiling .pio/build/BIGTREE_BTT002/src/src/lcd/marlinui.cpp.o
                 from Marlin/src/gcode/stats/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/stats/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/stats/../../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/stats/../gcode.h:303,
                 from Marlin/src/gcode/stats/M75-M78.cpp:23:
Marlin/src/gcode/stats/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/stats/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
Compiling .pio/build/BIGTREE_BTT002/src/src/lcd/menu/menu.cpp.o
Compiling .pio/build/BIGTREE_BTT002/src/src/lcd/menu/menu_advanced.cpp.o
Compiling .pio/build/BIGTREE_BTT002/src/src/lcd/menu/menu_bed_leveling.cpp.o
In file included from Marlin/src/gcode/temp/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/temp/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/temp/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/temp/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/temp/../../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/temp/../gcode.h:303,
                 from Marlin/src/gcode/temp/M104_M109.cpp:33:
Marlin/src/gcode/temp/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/temp/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
Compiling .pio/build/BIGTREE_BTT002/src/src/lcd/menu/menu_cancelobject.cpp.o
      |                                                                                                                       ^
In file included from Marlin/src/gcode/temp/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/temp/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/temp/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/temp/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/temp/../../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/temp/../gcode.h:303,
                 from Marlin/src/gcode/temp/M105.cpp:23:
Marlin/src/gcode/temp/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/temp/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
Compiling .pio/build/BIGTREE_BTT002/src/src/lcd/menu/menu_configuration.cpp.o
In file included from Marlin/src/gcode/temp/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/temp/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/temp/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/temp/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/temp/../../inc/MarlinConfig.h:31,
Compiling .pio/build/BIGTREE_BTT002/src/src/lcd/menu/menu_filament.cpp.o
                 from Marlin/src/gcode/temp/M106_M107.cpp:23:
Marlin/src/gcode/temp/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/temp/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
In file included from Marlin/src/gcode/temp/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/temp/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/temp/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/temp/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/temp/../../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/temp/M140_M190.cpp:29:
Marlin/src/gcode/temp/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/temp/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
In file included from Marlin/src/gcode/temp/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/temp/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/temp/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/temp/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/temp/../../inc/MarlinConfig.h:31,
Compiling .pio/build/BIGTREE_BTT002/src/src/lcd/menu/menu_game.cpp.o
                 from Marlin/src/gcode/temp/M141_M191.cpp:29:
Marlin/src/gcode/temp/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/temp/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
In file included from Marlin/src/gcode/temp/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
Compiling .pio/build/BIGTREE_BTT002/src/src/lcd/menu/menu_info.cpp.o
                 from Marlin/src/gcode/temp/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/temp/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/temp/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/temp/../../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/temp/M155.cpp:23:
Marlin/src/gcode/temp/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/temp/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
Compiling .pio/build/BIGTREE_BTT002/src/src/lcd/menu/menu_job_recovery.cpp.o
In file included from Marlin/src/gcode/temp/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/temp/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/temp/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/temp/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/temp/../../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/temp/M303.cpp:23:
Marlin/src/gcode/temp/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/temp/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
Compiling .pio/build/BIGTREE_BTT002/src/src/lcd/menu/menu_main.cpp.o
Compiling .pio/build/BIGTREE_BTT002/src/src/lcd/menu/menu_media.cpp.o
In file included from Marlin/src/gcode/units/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/gcode/units/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/gcode/units/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/gcode/units/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/gcode/units/../../inc/MarlinConfig.h:31,
                 from Marlin/src/gcode/units/../gcode.h:303,
                 from Marlin/src/gcode/units/M82_M83.cpp:23:
Marlin/src/gcode/units/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/gcode/units/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
In file included from Marlin/src/lcd/HD44780/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/lcd/HD44780/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/lcd/HD44780/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/lcd/HD44780/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/lcd/HD44780/../../inc/MarlinConfig.h:31,
                 from Marlin/src/lcd/HD44780/../marlinui.h:24,
                 from Marlin/src/lcd/HD44780/lcdprint_hd44780.cpp:19:
Marlin/src/lcd/HD44780/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/lcd/HD44780/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
      |                                                                                                                       ^
Compiling .pio/build/BIGTREE_BTT002/src/src/lcd/menu/menu_mixer.cpp.o
Compiling .pio/build/BIGTREE_BTT002/src/src/lcd/menu/menu_motion.cpp.o
In file included from .pio/libdeps/BIGTREE_BTT002/LiquidCrystal/LiquidCrystal.h:45,
                 from Marlin/src/lcd/HD44780/marlinui_HD44780.h:101,
                 from Marlin/src/lcd/HD44780/lcdprint_hd44780.cpp:22:
.pio/libdeps/BIGTREE_BTT002/LiquidCrystal/FastIO.h: At global scope:
.pio/libdeps/BIGTREE_BTT002/LiquidCrystal/FastIO.h:217:106: warning: 'boolean' is deprecated [-Wdeprecated-declarations]
  217 | void fio_shiftOut1(fio_register shift1Register, fio_bit shift1Bit, uint8_t value, boolean noLatch = false);
      |                                                                                                          ^
In file included from /Users/keith/.platformio/packages/framework-arduinoststm32/cores/arduino/wiring.h:34,
Compiling .pio/build/BIGTREE_BTT002/src/src/lcd/menu/menu_probe_offset.cpp.o
                 from /Users/keith/.platformio/packages/framework-arduinoststm32/cores/arduino/Arduino.h:36,
                 from Marlin/src/lcd/HD44780/../../inc/../HAL/./STM32/../shared/Marduino.h:36,
                 from Marlin/src/lcd/HD44780/../../inc/../HAL/./STM32/HAL.h:28,
                 from Marlin/src/lcd/HD44780/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/lcd/HD44780/../../inc/MarlinConfig.h:31,
                 from Marlin/src/lcd/HD44780/../marlinui.h:24,
                 from Marlin/src/lcd/HD44780/lcdprint_hd44780.cpp:19:
/Users/keith/.platformio/packages/framework-arduinoststm32/cores/arduino/wiring_constants.h:110:14: note: declared here
  110 | typedef bool boolean __attribute__((deprecated));
      |              ^~~~~~~
In file included from .pio/libdeps/BIGTREE_BTT002/LiquidCrystal/LiquidCrystal.h:45,
                 from Marlin/src/lcd/HD44780/marlinui_HD44780.h:101,
                 from Marlin/src/lcd/HD44780/lcdprint_hd44780.cpp:22:
.pio/libdeps/BIGTREE_BTT002/LiquidCrystal/FastIO.h:225:71: warning: 'boolean' is deprecated [-Wdeprecated-declarations]
  225 | void fio_shiftOut1(uint8_t pin, uint8_t value, boolean noLatch = false);
      |                                                                       ^
In file included from /Users/keith/.platformio/packages/framework-arduinoststm32/cores/arduino/wiring.h:34,
                 from /Users/keith/.platformio/packages/framework-arduinoststm32/cores/arduino/Arduino.h:36,
                 from Marlin/src/lcd/HD44780/../../inc/../HAL/./STM32/../shared/Marduino.h:36,
                 from Marlin/src/lcd/HD44780/../../inc/../HAL/./STM32/HAL.h:28,
                 from Marlin/src/lcd/HD44780/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/lcd/HD44780/../../inc/MarlinConfig.h:31,
                 from Marlin/src/lcd/HD44780/../marlinui.h:24,
                 from Marlin/src/lcd/HD44780/lcdprint_hd44780.cpp:19:
/Users/keith/.platformio/packages/framework-arduinoststm32/cores/arduino/wiring_constants.h:110:14: note: declared here
  110 | typedef bool boolean __attribute__((deprecated));
      |              ^~~~~~~
In file included from Marlin/src/lcd/HD44780/../../inc/../HAL/./STM32/../../core/serial_hook.h:24,
                 from Marlin/src/lcd/HD44780/../../inc/../HAL/./STM32/MarlinSerial.h:27,
                 from Marlin/src/lcd/HD44780/../../inc/../HAL/./STM32/HAL.h:34,
                 from Marlin/src/lcd/HD44780/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/lcd/HD44780/../../inc/MarlinConfig.h:31,
                 from Marlin/src/lcd/HD44780/marlinui_HD44780.h:28,
                 from Marlin/src/lcd/HD44780/marlinui_HD44780.cpp:34:
Compiling .pio/build/BIGTREE_BTT002/src/src/lcd/menu/menu_temperature.cpp.o
Marlin/src/lcd/HD44780/../../inc/../HAL/./STM32/../../core/serial_base.h: In member function 'bool SerialBase<Child>::has_feature(serial_index_t, SerialFeature) const':
Marlin/src/lcd/HD44780/../../inc/../HAL/./STM32/../../core/serial_base.h:127:119: warning: no return statement in function returning non-void [-Wreturn-type]
  127 |   bool has_feature(serial_index_t index, SerialFeature flag) const { (features(index) & flag) != SerialFeature::None; }
```

</p>
</details>
